src/groups/mqb/mqbc/mqbc_storagemanager.cpp

```diff
@@ -2379,6 +2418,68 @@ void StorageManager::do_processBufferedLiveData(const PartitionFSMArgsSp& args)
     }
 }
 
+void StorageManager::do_processBufferedPrimaryStatusAdvisories(
+    const PartitionFSMArgsSp& args)
+{
+    // executed by the *QUEUE DISPATCHER* thread associated with the paritionId
+    // contained in 'args'
+
+    // PRECONDITIONS
+    BSLS_ASSERT_SAFE(!args->eventsQueue()->empty());
+
+    const PartitionFSM::EventWithData& eventWithData =
+        args->eventsQueue()->front();
+    const EventData& eventDataVec = eventWithData.second;
+
+    BSLS_ASSERT_SAFE(eventDataVec.size() == 1);
+
+    const PartitionFSMEventData& eventData   = eventDataVec[0];
+    const int                    partitionId = eventData.partitionId();
+
+    if (d_cluster_p->isStopping()) {
+        BALL_LOG_WARN << d_clusterData_p->identity().description()
+                      << " Partition [" << partitionId << "]: "
+                      << "Cluster is stopping; skipping processing of "
+                      << "primary status advisory.";
+        return;  // RETURN
+    }
+
+    BALL_LOG_INFO
+        << d_clusterData_p->identity().description() << " Partition ["
+        << partitionId << "]: " << "Processing "
+        << d_bufferedPrimaryStatusAdvisoryInfosVec[partitionId].size()
+        << " buffered primary status advisory.";
+
+    for (PrimaryStatusAdvisoryInfosCIter cit =
+             d_bufferedPrimaryStatusAdvisoryInfosVec[partitionId].cbegin();
+         cit != d_bufferedPrimaryStatusAdvisoryInfosVec[partitionId].cend();
+         ++cit) {
+        BSLS_ASSERT_SAFE(cit->first.partitionId() == partitionId);
+
+        PartitionInfo& pinfo = d_partitionInfoVec[partitionId];
+        if (cit->second->nodeId() != pinfo.primary()->nodeId() ||
+            cit->first.primaryLeaseId() != pinfo.primaryLeaseId()) {
+            BALL_LOG_INFO << d_clusterData_p->identity().description()
+                          << " Partition [" << partitionId
+                          << "]: " << "Ignoring primary status advisory "
+                          << cit->first
+                          << " because primary node or leaseId is invalid. "
+                          << "Self-perceived [prmary, leaseId] is: ["
+                          << pinfo.primary()->nodeDescription() << ","
+                          << pinfo.primaryLeaseId() << "]";
+            continue;  // CONTINUE
+        }
+        pinfo.setPrimaryStatus(cit->first.status());
```

Comment:
What if you get a series of ACTIVE, then PASSIVE, then ACTIVE buffered advisories? How are you going to process them?

Reply:
Currently, we do not do anything special when a primary becomes ACTIVE -> PASSIVE -> ACTIVE. Thus, we also do no-op here. Arguably, we can improve the logic, but it's beyond the scope of this PR.